### ISSUE TITLE
Refactor test suites

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#436]
The way we write our test suites is inconsistent:

### `noUnusedVariableRuleTests.ts`

```ts
    it("restricts unused class members", () => {
        var fileName = "rules/nounusedvariable-class.test.ts";
        var Rule = Lint.Test.getRule("no-unused-variable");
        var failure1 = Lint.Test.createFailuresOnFile(fileName, Rule.FAILURE_STRING + "'z2'")([2, 13], [2, 15]);
        var failure2 = Lint.Test.createFailuresOnFile(fileName, Rule.FAILURE_STRING + "'mfunc4'")([18, 13], [18, 19]);

        var actualFailures = Lint.Test.applyRuleOnFile(fileName, Rule);

        assert.lengthOf(actualFailures, 2);
        Lint.Test.assertContainsFailure(actualFailures, failure1);
        Lint.Test.assertContainsFailure(actualFailures, failure2);
    });
```

### `noDuplicateKeyRuleTests.ts`

```ts
    it("forbids duplicate keys in object literals", () => {
        const fileName = "rules/dupkey.test.ts";
        const NoDuplicateKeyRule = Lint.Test.getRule("no-duplicate-key");
        const failureString = NoDuplicateKeyRule.FAILURE_STRING;

        const actualFailures = Lint.Test.applyRuleOnFile(fileName, NoDuplicateKeyRule);
        const createFailure1 = Lint.Test.createFailuresOnFile(fileName, failureString + "axa'");
        const createFailure2 = Lint.Test.createFailuresOnFile(fileName, failureString + "bd'");
        const createFailure3 = Lint.Test.createFailuresOnFile(fileName, failureString + "duplicated'");
        const expectedFailures = [
            createFailure1([10, 5], [10, 8]),
            createFailure2([13, 5], [13, 7]),
            createFailure1([14, 5], [14, 8]),
            createFailure3([31, 5], [31, 15])
        ];

        Lint.Test.assertFailuresEqual(actualFailures, expectedFailures);
    });
```

### `whitespaceRuleTests.ts`

```ts
    const createFailure = Lint.Test.createFailuresOnFile(fileName, WhitespaceRule.FAILURE_STRING);

    ...

    it("enforces whitespace in variable definitions", () => {
        const expectedFailures = [
            createFailure([11, 10], [11, 11]),
            createFailure([11, 11], [11, 12]),
            createFailure([13, 11], [13, 12])
        ];

        expectedFailures.forEach((failure) => {
            Lint.Test.assertContainsFailure(actualFailures, failure);
        });
    });
```

-------

I like the third style shown here, from `whitespaceRuleTests.ts`; it's the most concise. Now that we have TS 1.5 syntax, we should even be able to write

```ts
for (const failure of expectedFailures) {
    Lint.Test.assertContainsFailure(actualFailures, failure);
}
```

For cases where you need to create failures with different failure strings, you can do something like:

```ts
    const createFailure = (suffix) => Lint.Test.createFailuresOnFile(fileName, WhitespaceRule.FAILURE_STRING + suffix);

    ...

    it("enforces whitespace in variable definitions", () => {
        const expectedFailures = [
            createFailure("foo")([11, 10], [11, 11]),
            createFailure("bar")([11, 11], [11, 12]),
            createFailure("baz")([13, 11], [13, 12])
        ];

        for (const failure of expectedFailures) {
            Lint.Test.assertContainsFailure(actualFailures, failure);
        }
    });

```

**Comment by gscshoyru** [comment]
The only issue with the third kind (not sure if this is intentional or not) is that it's missing a length check -- we need to check that actual failures and expected failures are the same length, because otherwise our testing will miss false positives (failures in actual failures not in expected failures).

**Comment by adidahiya** [comment]
yeah, you're right, there needs to be a line somewhere next to where the `actualFailures` are created to check length of the list.

maybe we should do it in a slightly more automated way?

```ts
const actualFailures = Lint.applyRuleOnFile(Rule, fileName);
const allExpectedFailures = [];

...

it("...", () => {
    const expectedFailures = [ ... ];
    allExpectedFailures.concat(expectedFailures);
    ...
});

...

it("doesn't have extra failures", () => {
    assert.lengthOf(actualFailures, allExpectedFailures.length);
});
```

**Comment by gscshoyru** [comment]
I see what you're saying, but tests shouldn't be expected to run in order like that; maintaining state between tests like that is not convention/good practice, since if you want to skip some of them or run them independently, they may break. 

I'm not sure a good way of doing it other than either having each test use a separate file (as some rules currently do) or manually counting yourself, though.

**Comment by adidahiya** [comment]
Ok; I'm fine with just using a static number value for length checks for now. We can refactor again if we come up with a better solution.
